File: implementations/python/d_ary_heap.py
"""
Reference Implementation: Contiguous Array-Backed d-ary Heap
Demonstrates arbitrary branching factor d (e.g., d = 4, d = 8),
generalized 0-based index arithmetic, sift-up (O(log_d n)),
sift-down with d-child scan (O(d log_d n)), and Floyd-style O(n) build_heap.

Language: Python 3
"""

from typing import Any, List, Callable, Optional
import unittest


class DAryHeap:
    def __init__(self, d: int = 4, elements: Optional[List[Any]] = None, key: Optional[Callable[[Any], Any]] = None):
        if d < 2:
            raise ValueError("Branching factor d must be at least 2")
        self._d = d
        self._key = key if key is not None else (lambda x: x)
        if elements is not None:
            self._data = list(elements)
            self._build_heap()
        else:
            self._data = []

    def __len__(self) -> int:
        return len(self._data)

    def is_empty(self) -> bool:
        return len(self._data) == 0

    def _parent(self, i: int) -> int:
        return (i - 1) // self._d

    def _first_child(self, i: int) -> int:
        return self._d * i + 1

    def _sift_up(self, i: int) -> None:
        while i > 0:
            p = self._parent(i)
            if self._key(self._data[i]) < self._key(self._data[p]):
                self._data[i], self._data[p] = self._data[p], self._data[i]
                i = p
            else:
                break

    def _sift_down(self, i: int, n: int) -> None:
        while True:
            best = i
            first = self._first_child(i)
            if first >= n:
                break

            last = min(first + self._d, n)
            for c in range(first, last):
                if self._key(self._data[c]) < self._key(self._data[best]):
                    best = c

            if best != i:
                self._data[i], self._data[best] = self._data[best], self._data[i]
                i = best
            else:
                break

    def _build_heap(self) -> None:
        n = len(self._data)
        if n > 1:
            for i in range((n - 2) // self._d, -1, -1):
                self._sift_down(i, n)

    def push(self, val: Any) -> None:
        self._data.append(val)
        self._sift_up(len(self._data) - 1)

    def pop(self) -> Any:
        if not self._data:
            raise IndexError("pop from an empty heap")
        top_val = self._data[0]
        last = self._data.pop()
        if self._data:
            self._data[0] = last
            self._sift_down(0, len(self._data))
        return top_val

    def peek(self) -> Any:
        if not self._data:
            raise IndexError("peek from an empty heap")
        return self._data[0]

    def is_valid_heap(self) -> bool:
        n = len(self._data)
        for i in range(n):
            first = self._first_child(i)
            last = min(first + self._d, n)
            for c in range(first, last):
                if self._key(self._data[c]) < self._key(self._data[i]):
                    return False
        return True


class TestDAryHeap(unittest.TestCase):
    def test_4ary_min_heap(self):
        heap = DAryHeap(d=4)
        self.assertTrue(heap.is_empty())
        self.assertEqual(len(heap), 0)

        inputs = [55, 12, 89, 4, 32, 1, 67, 23, 19, 78, 3, 99, 45, 6]
        for x in inputs:
            heap.push(x)
            self.assertTrue(heap.is_valid_heap())

        self.assertEqual(len(heap), len(inputs))
        self.assertEqual(heap.peek(), 1)

        sorted_out = []
        while not heap.is_empty():
            sorted_out.append(heap.pop())
            self.assertTrue(heap.is_valid_heap())

        self.assertEqual(sorted_out, sorted(inputs))

    def test_8ary_build_heap(self):
        inputs = [55, 12, 89, 4, 32, 1, 67, 23, 19, 78, 3, 99, 45, 6]
        heap8 = DAryHeap(d=8, elements=inputs)
        self.assertTrue(heap8.is_valid_heap())
        self.assertEqual(heap8.peek(), 1)

    def test_3ary_heap(self):
        inputs = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
        heap3 = DAryHeap(d=3, elements=inputs)
        self.assertTrue(heap3.is_valid_heap())
        self.assertEqual(heap3.peek(), 1)


if __name__ == "__main__":
    unittest.main()

File: implementations/cpp/d_ary_heap.cpp
/**
 * Reference Implementation: Contiguous Array-Backed d-ary Heap
 * Demonstrates arbitrary branching factor D (e.g., D = 4, D = 8),
 * generalized 0-based index arithmetic, sift-up (O(log_D n)),
 * sift-down with D-child scan (O(D log_D n)), and Floyd-style O(n) build_heap.
 *
 * Language: C++17
 */

#include <iostream>
#include <vector>
#include <functional>
#include <cassert>
#include <stdexcept>
#include <utility>
#include <algorithm>

template <typename T, std::size_t D = 4, typename Compare = std::less<T>>
class DAryHeap {
    static_assert(D >= 2, "d-ary heap branching factor D must be at least 2");

private:
    std::vector<T> data_;
    Compare comp_; // Returns true if a has higher priority than b

    static inline std::size_t parent(std::size_t i) noexcept {
        return (i - 1) / D;
    }

    static inline std::size_t first_child(std::size_t i) noexcept {
        return D * i + 1;
    }

    void sift_up(std::size_t i) {
        while (i > 0) {
            std::size_t p = parent(i);
            if (comp_(data_[i], data_[p])) {
                std::swap(data_[i], data_[p]);
                i = p;
            } else {
                break;
            }
        }
    }

    void sift_down(std::size_t i, std::size_t n) {
        while (true) {
            std::size_t best = i;
            std::size_t first = first_child(i);

            if (first >= n) {
                break; // No children
            }

            std::size_t last = std::min(first + D, n);
            for (std::size_t c = first; c < last; ++c) {
                if (comp_(data_[c], data_[best])) {
                    best = c;
                }
            }

            if (best != i) {
                std::swap(data_[i], data_[best]);
                i = best;
            } else {
                break;
            }
        }
    }

public:
    explicit DAryHeap(Compare comp = Compare{}) : comp_(comp) {}

    // Floyd-style O(n) bottom-up heap construction
    explicit DAryHeap(std::vector<T> elements, Compare comp = Compare{})
        : data_(std::move(elements)), comp_(comp) {
        if (data_.size() > 1) {
            for (std::size_t i = (data_.size() - 2) / D + 1; i > 0; --i) {
                sift_down(i - 1, data_.size());
            }
        }
    }

    [[nodiscard]] std::size_t size() const noexcept {
        return data_.size();
    }

    [[nodiscard]] bool empty() const noexcept {
        return data_.empty();
    }

    const T& top() const {
        if (data_.empty()) {
            throw std::underflow_error("DAryHeap::top(): heap is empty");
        }
        return data_[0];
    }

    void push(const T& val) {
        data_.push_back(val);
        sift_up(data_.size() - 1);
    }

    void push(T&& val) {
        data_.push_back(std::move(val));
        sift_up(data_.size() - 1);
    }

    T pop() {
        if (data_.empty()) {
            throw std::underflow_error("DAryHeap::pop(): heap is empty");
        }
        T top_val = std::move(data_[0]);
        data_[0] = std::move(data_.back());
        data_.pop_back();
        if (!data_.empty()) {
            sift_down(0, data_.size());
        }
        return top_val;
    }

    [[nodiscard]] bool is_valid_heap() const {
        for (std::size_t i = 0; i < data_.size(); ++i) {
            std::size_t first = first_child(i);
            std::size_t last = std::min(first + D, data_.size());
            for (std::size_t c = first; c < last; ++c) {
                if (comp_(data_[c], data_[i])) {
                    return false;
                }
            }
        }
        return true;
    }
};

int main() {
    // 1. 4-ary Min-Heap verification
    DAryHeap<int, 4> heap4;
    assert(heap4.empty());
    assert(heap4.size() == 0);

    std::vector<int> inputs = {55, 12, 89, 4, 32, 1, 67, 23, 19, 78, 3, 99, 45, 6};
    for (int x : inputs) {
        heap4.push(x);
        assert(heap4.is_valid_heap());
    }

    assert(heap4.size() == inputs.size());
    assert(heap4.top() == 1);

    std::vector<int> sorted4;
    while (!heap4.empty()) {
        sorted4.push_back(heap4.pop());
        assert(heap4.is_valid_heap());
    }
    assert(std::is_sorted(sorted4.begin(), sorted4.end()));

    // 2. 8-ary Max-Heap verification
    DAryHeap<int, 8, std::greater<int>> heap8;
    for (int x : inputs) {
        heap8.push(x);
        assert(heap8.is_valid_heap());
    }
    assert(heap8.top() == 99);

    std::vector<int> sorted8;
    while (!heap8.empty()) {
        sorted8.push_back(heap8.pop());
        assert(heap8.is_valid_heap());
    }
    assert(std::is_sorted(sorted8.rbegin(), sorted8.rend()));

    // 3. Floyd-style O(n) construction for D = 4
    DAryHeap<int, 4> floyd_heap4(inputs);
    assert(floyd_heap4.is_valid_heap());
    assert(floyd_heap4.top() == 1);

    // 4. Floyd-style O(n) construction for D = 3
    DAryHeap<int, 3> floyd_heap3(inputs);
    assert(floyd_heap3.is_valid_heap());
    assert(floyd_heap3.top() == 1);

    std::cout << "[PASS] All d-ary Heap C++ unit tests passed.\n";
    return 0;
}
